File: frontend/src/lib/inboundOrder.ts
// 入库单前端类型 + API。对齐后端 schemas/inbound_order.py。
// 🔴 入库单据/行**无任何成本字段**(契约 D3)—— 读投影天然无红线,前端不渲染任何价格列。
// 详情内嵌 PO 摘要的成本由后端按 purchase:read_cost 脱敏为 null;payable 块仅在响应含该键时渲染。
import { api } from "./api";
import type { Page } from "./catalog";
import type { PurchaseOrderOut } from "./purchaseOrder";
import type { PayableOut } from "./payable";
import { qs } from "./qs";

export type InboundOrderStatus = "IN_TRANSIT" | "RECEIVED" | "CANCELLED";

export interface InboundOrderOut {
  id: number;
  no: string;
  purchase_order_id: number;
  status: InboundOrderStatus;
  carrier_name: string | null;
  tracking_no: string | null;
  shipped_at: string | null;
  eta: string | null;
  arrived_at: string | null;
  remark: string | null;
  updated_at: string;
  purchase_order_no: string;
  supplier_display: string;
}

export interface InboundOrderLineOut {
  id: number;
  inbound_order_id: number;
  purchase_order_line_id: number;
  sku_id: number;
  name_snapshot: string;
  spec_text_snapshot: string;
  unit_snapshot: string;
  language: string;
  qty: number | string;
  sort_order: number;
  remark: string | null;
}

export interface InboundOrderListItem {
  id: number;
  no: string;
  purchase_order_id: number;
  purchase_order_no: string;
  supplier_display: string;
  status: InboundOrderStatus;
  carrier_name: string | null;
  tracking_no: string | null;
  eta: string | null;
  arrived_at: string | null;
  line_count: number;
  total_qty: number | string;
  created_at: string;
}

/** 建单器数据源(GET /inbound-orders/receivable-lines)。无成本字段。 */
export interface ReceivableLine {
  purchase_order_line_id: number;
  sku_id: number;
  name_snapshot: string;
  spec_text_snapshot: string;
  unit_snapshot: string;
  language: string;
  ordered_qty: number | string;
  in_transit_qty: number | string;
  received_qty: number | string;
  remaining_qty: number | string;
}

/** 入库详情返回(payable 仅在持 payable:read 且已入库时存在;前端按键存在与否渲染)。 */
export interface InboundOrderDetail {
  order: InboundOrderOut;
  lines: InboundOrderLineOut[];
  purchase_order: PurchaseOrderOut;
  payable?: PayableOut;
}

/** 写入行(引用 PO 行 + 本次入库数量)。 */
export interface InboundOrderLineIn {
  purchase_order_line_id: number;
  qty: number;
  remark?: string | null;
  sort_order?: number;
}

export interface InboundOrderCreateBody {
  purchase_order_id: number;
  carrier_name?: string | null;
  tracking_no?: string | null;
  shipped_at?: string | null;
  eta?: string | null;
  remark?: string | null;
  lines: InboundOrderLineIn[];
}

export interface InboundOrderUpdateBody {
  carrier_name?: string | null;
  tracking_no?: string | null;
  shipped_at?: string | null;
  eta?: string | null;
  remark?: string | null;
  lines: InboundOrderLineIn[];
  /** 乐观锁基线 = 打开编辑时的 order.updated_at(对齐 PO)。 */
  expected_updated_at: string;
}

export interface InboundOrderListFilters {
  status?: string;
  purchase_order_id?: number;
  supplier_id?: number;
  keyword?: string;
  page?: number;
  size?: number;
}

export const inboundOrderApi = {
  list: (p: InboundOrderListFilters) =>
    api.get<Page<InboundOrderListItem>>(`/api/v1/inbound-orders${qs(p as Record<string, unknown>)}`),
  get: (id: number) => api.get<InboundOrderDetail>(`/api/v1/inbound-orders/${id}`),
  receivableLines: (purchaseOrderId: number) =>
    api.get<{ items: ReceivableLine[] }>(
      `/api/v1/inbound-orders/receivable-lines?purchase_order_id=${purchaseOrderId}`,
    ),
  create: (b: InboundOrderCreateBody) =>
    api.post<InboundOrderDetail>("/api/v1/inbound-orders", b),
  update: (id: number, b: InboundOrderUpdateBody) =>
    api.put<InboundOrderDetail>(`/api/v1/inbound-orders/${id}`, b),
  receive: (id: number, arrived_at?: string | null) =>
    api.post<InboundOrderDetail>(`/api/v1/inbound-orders/${id}/receive`, { arrived_at: arrived_at ?? null }),
  unreceive: (id: number, void_reason?: string | null) =>
    api.post<InboundOrderDetail>(`/api/v1/inbound-orders/${id}/unreceive`, { void_reason: void_reason ?? null }),
  cancel: (id: number) => api.post<InboundOrderDetail>(`/api/v1/inbound-orders/${id}/cancel`),
};

File: backend/app/schemas/inbound_order.py
"""入库单 schemas。

🔴红线由设计消除:入库单/行**无任何成本字段**(契约 D3),故读投影无需脱敏工厂。
详情内嵌的 PO 摘要走既有 PurchaseOrderOut.build(can_see_cost=...) 脱敏;
payable 块仅在持 payable:read 时由端点组装(整块门控,非字段级)。
"""
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, Field

# ---------- 写入(建单 / 整单保存)----------


class InboundOrderLineIn(BaseModel):
    """入库行:引用 PO 行 + 本次入库数量。无成本字段(平移快照在 service 从 PO 行取)。"""
    purchase_order_line_id: int
    qty: float = Field(gt=0)
    remark: str | None = None
    sort_order: int = Field(default=0, ge=0)


class InboundOrderCreateIn(BaseModel):
    """基于 CONFIRMED PO 建一张在途入库单。lines 空 → service 抛 41707。"""
    purchase_order_id: int
    carrier_name: str | None = None
    tracking_no: str | None = None
    shipped_at: date | None = None
    eta: date | None = None
    remark: str | None = None
    lines: list[InboundOrderLineIn] = Field(default_factory=list)


class InboundOrderUpdateIn(BaseModel):
    """整单保存(PUT):仅 IN_TRANSIT;必带乐观锁基线(对齐 PO)。PO 归属不可改(单据身份)。行整表重写。"""
    carrier_name: str | None = None
    tracking_no: str | None = None
    shipped_at: date | None = None
    eta: date | None = None
    remark: str | None = None
    lines: list[InboundOrderLineIn] = Field(default_factory=list)
    expected_updated_at: datetime


class InboundOrderReceiveIn(BaseModel):
    """确认入库:实际到货日(默认当天,service 填)。"""
    arrived_at: date | None = None


class InboundOrderUnreceiveIn(BaseModel):
    """撤销入库:留痕原因(可空)。"""
    void_reason: str | None = None


# ---------- 读投影(无成本字段)----------


class InboundOrderLineOut(BaseModel):
    id: int
    inbound_order_id: int
    purchase_order_line_id: int
    sku_id: int
    name_snapshot: str
    spec_text_snapshot: str
    unit_snapshot: str
    language: str
    qty: float
    sort_order: int
    remark: str | None

    @classmethod
    def build(cls, line) -> dict:
        return cls.model_validate(line, from_attributes=True).model_dump()


class InboundOrderOut(BaseModel):
    id: int
    no: str
    purchase_order_id: int
    status: str
    carrier_name: str | None
    tracking_no: str | None
    shipped_at: date | None
    eta: date | None
    arrived_at: date | None
    remark: str | None
    updated_at: datetime

    @classmethod
    def build(cls, order, extra: dict | None = None) -> dict:
        d = cls.model_validate(order, from_attributes=True).model_dump()
        if extra:
            d.update(extra)
        return d


class InboundOrderListItem(BaseModel):
    id: int
    no: str
    purchase_order_id: int
    purchase_order_no: str
    supplier_display: str
    status: str
    carrier_name: str | None
    tracking_no: str | None
    eta: date | None
    arrived_at: date | None
    line_count: int
    total_qty: float
    created_at: datetime


class ReceivableLineOut(BaseModel):
    """PO 可收行(建单器数据源)。无成本字段。"""
    purchase_order_line_id: int
    sku_id: int
    name_snapshot: str
    spec_text_snapshot: str
    unit_snapshot: str
    language: str
    ordered_qty: float
    in_transit_qty: float
    received_qty: float
    remaining_qty: float


class RelatedInboundOrderItem(BaseModel):
    """PO 详情「入库记录区」项(含 CANCELLED 并标状态,可追溯)。无成本字段。"""
    id: int
    no: str
    status: str
    carrier_name: str | None
    tracking_no: str | None
    eta: date | None
    arrived_at: date | None
    created_at: datetime
